README
Snack v2.2.11 - a sound toolkit for scripting languages

Update by Scott Pitcher, 31OCT2021.

This archive contains an updated version of the Snack sounds library.
I've incorporated some bug fixes which included correcting build errors
when building with MINGW32 or 64, some SEGV errors have been corrected
and I've incorporated at least one bugfix that was emailed to me.

I'm putting this archive on Github so (1) binary version for Windows
are readily available and (2) the source can be maintained in one place.

Now, before I put this source archive together I did email both KTH and
KTH Division of Speech, Music and Hearing looking for someone who either
maintains Snack or knows anything about it, but the only reply I received
was from info@kth.se suggesting I email the second group, but I didn't
receive any reply from them. So I've included the few old source versions
I was able to find for completeness, and I've built this 2.2.11 ontop of
2.2.10.

I've uploaded binary builds for Windows 32 and 64.

Please email me if you have any trouble with these.
Please email me if you have any other patches I should include.


---------------------------------------------------------------------
Snack v2.2 - a sound toolkit for scripting languages

Kåre Sjölander

Email address:
There are several anti-spam methods between me and anybody trying to contact
me by e-mail. If you don't hear anything try to contact me again. One option
is also to use this forum
http://www1.speech.kth.se/prod/wavesurferforum/phpBB2/index.php
which can be used to discuss Snack issues.
Also, please start the subject line with the word "snack" in any e-mail
correspondence in order to clearly indicate that it is an important message.
<xxxxxxxxxxxxxxxxxx>


Introduction
-------------------------------------------------------------------

Snack is a sound processing toolkit designed as an extension
to a scripting language. Snack currently works with the scripting
languages Tcl/Tk, Python, and Ruby.
Snack has commands to play, record, process, and visualize sound.
Snack provides high level sound objects, with flexible storage management,
and streaming support. It handles most common sound file formats.
The visualization canvas item types update in real-time and can output
postscript. The same scripts run on Unix (Linux, Solaris, HP-UX, IRIX,
FreeBSD, NetBSD), Macintosh, and Windows 95/98/NT/2000/XP.
Several example scripts can be found in the demos directory.

For more information on Tcl/Tk and Python see
Tcl/Tk: http://tcl.activestate.com/
Python: http://www.python.org/

In order to use Snack you must have Tcl and Tk version 8.0 or later. Always 
use the latest stable release, currently 8.4.8, which can be downloaded from 
ftp.activestate.com/ActiveTcl/html/.

Tcl/Tk is often included with Python distributions. Snack has been tested
with Python 1.5.2 - 2.3 on Linux and Windows. See the Python
section below.

The Snack package has been tested on the following platforms:
Windows 95/98/NT/2000/XP
Linux 2.0.35 (OSS 3.5) - 2.4.20-31.9 (OSS 3.8.2/ALSA 0.9.0) (x86 & PowerPC)
SunOS 5.5.1 (Sparc and x86), Solaris 8 (Sparc)
HP-UX B.10.20/11.00
IRIX 6.2, 6.3 & 6.4 (even runs under SoftWindows, although slowly)
NetBSD-current/i386 using the OSS emulation library
BSD/OS-4.1
Mac OS X, MacOS 8.6 and later

Make sure you have the latest Snack version available at
http://www.speech.kth.se/snack/ before proceeding with the installation.



Compilation instructions for Snack on Unix (see mac/README for OS X)
-------------------------------------------------------------------

You must have Tcl/Tk installed on your system. Both Tcl and Tk must have
been configured with the "--enable-shared" flag at build time (this is 
default for the 8.4 versions). Build them before Snack.

Unpack Snack and change directory to the snack2.2.10/unix directory.

Type "./configure".

If you get this error message "Can't find Tcl configuration definitions",
you have to specify their locations using the options "--with-tcl" and
"--with-tk" when running configure. For example, use 
"--with-tcl=/usr/lib --with-tk=/usr/lib" if Tcl/Tk was
installed in "/usr/lib", or "--with-tcl=../tcl8.4.8/unix" if you use that
version.

Linux users who have the ALSA sound driver installed should specify the configure flag --enable-alsa in order to enable native ALSA support.

Type "make".

If the make fails you will have to edit the Makefile and possibly also the 
configure script. TCL_INCPATH should be the path to the directory which 
contains the Tcl header file (tcl.h). The same goes for TK_INCPATH (tk.h). 
XINCLUDES should be the path to the directory containing the X11 include
files.

If you use Tcl/Tk 8.0 to 8.0.5 you will have to use the flag
--disable-stubs, when configuring Snack.

Summary in case you download Tcl, Tk, and Snack and build from scratch:
You should place the distributions in the same directory and unpack them.
Build Snack using the following commands:
cd tcl8.4.8/unix
./configure --enable-shared
cd ../../tk8.4.8/unix
./configure --enable-shared
cd ../../snack2.2.10/unix
./configure
make


Testing
-------------------------------------------------------------------

When you have succesfully compiled Snack it's time to look at the demos.
In order to run these you will have to set the environment variable 
TCLLIBPATH to the directory containing the Snack library files (you don't 
need this step if you install Snack as described below). If you have 
compiled Snack in /u/kare/snack2.2.10/unix, this would be:

setenv TCLLIBPATH /u/kare/snack2.2.10/unix

Change to the demos/tcl/ directory and type ./widget.tcl:

cd ../demos/tcl/
./widget.tcl

this runs the Snack widget demonstration.

You should also try the Snack test suite. Type:

make test

in the build directory. You should then see a printout of the test files
processed. If any errors occur, you'll see a much more substantial printout
for each error. Please report any problems to the author.



System wide installation
-------------------------------------------------------------------

The recommended installation directory is the standard Tcl package directory. 
You can determine this by typing:

tclsh
puts $tcl_pkgPath

If you choose this directory you don't have to set the TCLLIBPATH variable.
Edit the Makefile if necessary. SNACK_INSTALL_PATH should be the path to the 
directory where you want to install the Snack package (e.g. /usr/local/lib).

Install using:

make install



Compilation instructions for Snack on Windows
-------------------------------------------------------------------

The MSVC++ 6.0 workspace and project files necessary to build Snack on Windows
are located in the win directory of the source distribution.

In order to use Snack you will have to install Tcl/Tk 8.4 first.

Snack's project files are set up for use with Tcl installed in C:\Tcl\, it
might be easiest to change such paths by editing all *.dsp files in a text
editor.
Don't forget to copy the files snack.tcl and pkgIndex.tcl.dll from the
unix directory to C:\Tcl\lib\snack2.2\ and rename the last file pkgIndex.tcl


Compilation instructions for Snack on Windows using Mingw
-------------------------------------------------------------------

Mingw users (gcc for Windows) can now compile Snack under Windows.

The win subdirectory contains a configure script that can be
used to compile snack with Mingw compiled versions of Tcl and Tk.
Users will first need to download Tcl and Tk sources along with
the msys_mingw package from the following URL:

http://sourceforge.net/project/showfiles.php?group_id=10894

At this point, the DirectX 6 headers and lib files will need
to be downloaded and installed into /mingw/dx6. If you change
the location of the DirectX install, the Makefile will need
to be updated after configure is run.

http://people.montana.com/%7Ebowman/Software/ming.zip

To build, run the configure script and pass the --with-tcl=DIR and
--with-tk=DIR arguments that correspond to the build directories
for Tcl and Tk. Tcl/Tk 8.4 or newer is required, OGG and NIST
format options are not currently supported with the Mingw build.

The typical configure-make cycle looks like:
./configure --with-tcl=$TCL_DIR --with-tk=$TK_DIR --prefix=$INST_DIR
make
make install

The above will create a snack directory in $INST_DIR/lib, which should
be the same --prefix that you passed to the Tcl and Tk configure scripts.

Using Snack with Python
-------------------------------------------------------------------

Snack must currently be used in conjunction with Tkinter. You need to have a
matching installation of Snack and Tcl. For example, Python 1.5.2 uses
Tcl8.0.5 and Python 1.6-2.3 uses Tcl8.3 on Windows. For Unix you have to
edit Modules/Setup before building Python to specify which Tcl version to use.
The default is Tcl8.0. You can check this by typing the following lines
at the Python prompt:

import Tkinter
Tkinter.Tk().tk.eval('info tclversion')

Snack is accessed using the tkSnack module located in the python directory.
You need to put the file tkSnack.py in your Python path. 

Some demos can be found in demos/python/.

Documentation is found in doc/python-man.html



Using Snack with NIST/Sphere file formats
-------------------------------------------------------------------

It is possible to configure Snack to create an additional package for
handling the NIST/Sphere file formats. Use the configure options
--with-nist-include and --with-nist-lib to specify the locations of
the corresponding directories of the NIST/Sphere distribution,
which can be downloaded at
ftp://jaguar.ncsl.nist.gov/pub/sphere_2.6a.tar.Z. In the Sphere
distribution you will have to modify the file src/scripts/install.sh and add
the compilation flag to create position independent code in some cases. Look
at the Snack Makefile for these.

On HP-UX with HP cc use: -Ae +z
Linux users should specify platform: 10 (Custom),
compiler flags: -O -Wall -fPIC, architecture: INTEL,and comment the
line "extern char *sys_errlist[];" in nist/src/lib/sp/exit.c
See SphereFile.txt for more info.

Windows users can get a compiled dll at the address
http://www.speech.kth.se/~kare/libsnacksphere.dll
Put it in ..Tcl\lib\Snack2.2 (together with libsnack.dll).

Linux users get a compiled library at the address
http://www.speech.kth.se/~kare/libsnacksphere.so
Put it in the Snack installation directory (together with libsnack.so).



Using Snack with the OGG/Vorbis file format
-------------------------------------------------------------------

It is possible to configure Snack to create an additional package for
handling the OGG/Vorbis file format. Use the configure options
--with-ogg-include and --with-ogg-lib to specify the locations of an
existing OGG/Vorbis installation.
More information about OGG/Vorbis available at http://www.vorbis.com/

Windows users can get a compiled dll at the address
http://www.speech.kth.se/~kare/ogg.tar.gz
Put libsnackogg.dll in ..Tcl\lib\Snack2.2 (together with libsnack.dll).

Linux users can get a compiled library at the address
http://www.speech.kth.se/~kare/ogg.tar.gz
Put libsnackogg.so in the Snack installation directory (together 
with libsnack.so).

When building Ogg/Vorbis from source on Unix note that some modifications
to the Makefiles might be necessary. Specify
"AM_CPPFLAGS=-I$prefix/include" "LDFLAGS=-L$exec_prefix/lib"
to get them to find their own installed headers and libraries.
On Solaris force libvorbis to use the native compiler and add
-Dalloca=__builtin_alloca
to prevent the system from using a non-existing alloca.



Dynamic linking problems
-------------------------------------------------------------------

(Sun, Linux, and SGI)
The loader must be able to find the "libsnack.so" shared object. You may
have to set the environment variable LD_LIBRARY_PATH to point the directory
where it is located. This variable is a colon separated list of paths which
can be modified using setenv. If you, for example, have compiled Snack in
the directory /u/kare/snack/unix and have the shared objects libtcl8.4.so
and libtk8.4.so in /usr/local/bin you would use the following command at the 
terminal prompt or in your .cshrc file:

setenv LD_LIBRARY_PATH /u/kare/snack/unix:/usr/local/bin:$LD_LIBRARY_PATH

(HP)
The corresponding environment variable on HP-UX systems is SHLIB_PATH and 
shared objects use ".sl" as file extension.


License
-------------------------------------------------------------------

Snack is currently licensed under GPL, see the file COPYING.
Contact the author if this kind of license is a problem.
The only reason for GPL is the MPEG decoding code, see jkFormatMP3.c.
If you remove that code it is possible to use a license similar to the one 
Tcl/Tk uses (BSD style) and to include Snack in any software on the same
terms.



Acknowledgements
-------------------------------------------------------------------
The following people have contributed code, suggestions, and/or other help.

Roger Lindell
Jonas Beskow
Claude Barras
Mel
Alexandre Ferrieux
Daryl Vertigan
David van Leeuwen
Stefan Karnebäck
Simon Arnfield
Magnus Lundeberg
Zhibiao Wu
Qiru Zhou
Philippe Langlais
Dan Ellis
Holger Kanwischer
Egbert Ammicht
Richard Donselius
Seymour Shlien
Leonid Spektor
Frederic Bonnet
Andreas Gustafsson
Jeffrey Hobbs
Johan Dahl
Erhard Rank
Richard Ross-Langley
Uwe Koloska
Kevin Russell
Nicolas Garbe
Tom Wilkason
Jonas Ekeroot
Peter Kabal
Ted Dunning
Andreas Kupries
Erik Allaert
Christian Wegehaupt
Alex Barnes
Steve Davis
Oriol Pique
Athanassios Hatzis
Scott McSpadden
Albert Visagie
Helmut Hissen
Daniel A. Steffen
Hans Werner Strube
David McClosky
Russell Owen
Benjamin Baudoux
Steve Renals
Michel Scheffers
Reinhard Max
Brian Theado
Håkan Melin
Scott T. Silliman
Mo DeJong
Richard Dobson
Alan Schnarr
Lorenzo Seno
Edouard Hinard


changes
Changes to Snack

----------------- Released 1.3, 10/1/98 -----------------------

10/1/98 (new feature) AIFF file support, although there seems to be lots of
files out there that don't adhere to the specification.

10/1/98 (new feature) Sound objects now can be linked to disk files and
streams (Tcl channels) as well as having their audio data in memory.

10/5/98 (new feature) Added -byteOrder option to the sound data command and to all commands taking the -byteswap option, which is deprecated.

10/7/98 (bug fix) Fixed problem playing zero-length sounds on HP-UX.

10/7/98 (bug fix) Fixed problem playing multi-channel disk files.

10/12/98 (bug fix) The configure command now stops any active play/record operation before setting an option. Also, corrected -format and -channels options.

10/12/98 (new feature) Added cget command to complement configure command.

10/14/98 (bug fix) Fixed problem with insert command.

10/21/98 (bug fix) Fixed problem with APause() on IRIX.

----------------- Released 1.4b, 10/26/98 -----------------------

10/26/98 (new feature) Makefile now also builds a Tcl only libsound.so

10/28/98 (new feature) Split OSound.c into jkSound*.c and added Snack_AddSoundCmd() for Snack extensions. Added pause command.

10/30/98 (bug fix) StopSound() now only called if the format of the sound samples change while a sound is playing, as a result of read or copy, e.g.

11/3/98 (bug fix) Fixed bug in destroyCmd() and flushCmd() causing unwanted closing of audio device.

11/6/98 (bug fix) Fixed bug in Snack_SoundCmd() for -load'ing raw files.

11/11/98 (new feature) Better validation of options -fftlen/-winlen/-channel.

11/18/98 (bug fix) Snack now more robust to audio device peculiarities, misc audio fixes for HP-UX, Solaris, Linux, and Windows.

11/18/98 (new feature) Play commands now queue up and play in sequence.

11/22/98 (bug fix) Fixed bugs in SmpMov() and lengthCmd() for byte sized sample data.

12/19/98 (new feature) Added -colormap option to spectrograms.

12/20/98 (new feature) Added on-the-fly sample conversion for playback, Lin8 on HP-UX and u-/A-law on Linux (if not supported by the hardware).  

12/22/98 (new feature) PS generation for color spectrograms.

----------------- Released 1.4p2, 1/1/99 -----------------------

1/1/99 (new feature) Added -subsample option to waveforms. Now large sounds can be sub-sampled when drawing the waveform envelope. This can be much faster. For some synthetic sounds this can be problematic, in this case setting -subsample to 1 gives the old behaviour, i.e. don't sub-sample. Fixed drawing of zoomed sounds which look nicer now.

1/1/99 (new feature) Made some optimizations to spectrogram drawing. Added -gridcolor option. 256-color displays now supported under Windows.

1/4/99 (new feature) Raw file input now tries to guess byte order and sample format if they weren't specified in the read/data commands.

1/7/99 (new feature) Added -fileformat option to the sound write command which overrides the guess of format from the file name extension.

1/15/99 (new feature) Added sound convert command. Rather simple resampling algorithm (to start with).

1/26/99 (new feature) Modified spectrogram canvas item code with the goal of making it replace the spectrogram widget, which is now deprecated.

1/27/99 (bug fix) Swapped Lin8 and Lin8Offset everywhere, long overdue.

1/29/99 (new feature) Split the file AudIO.c into one jkAudIO_*.c per platform. Renamed some other source and header files too.

2/1/99 (bug fix) Waveforms redraw properly now.

----------------- Released 1.5a1, 2/8/99 -----------------------

2/8/99 (new feature) It is now possible to add new sound file formats using the Snack_AddFileFormat() function. Only reading supported to start with.

2/8/99 (new feature) Added support for the NIST/Sphere file formats (SphereFile.c). This must be specified at configure time, see the README file.

2/8/99 (new feature) Automatic detection of byte order and sample encoding format now also for -file/-channel sounds.

2/8/99 (new feature) Snack now uses immediate loading. This eliminates the problems with, for example, finding the Snack item types in some cases.

2/8/99 (feature change) Waveform -subsample option now has a default value of 1.

2/9/99 (bug fix) Copy command now allocates sufficient memory when copying multi-channel sounds.

2/10/99 (bug fix) Fixed canvas item bug when UpdateXXX() is called before an item has been drawn on screen.

----------------- Released 1.5a2, 2/10/99 -----------------------

2/12/99 (bug fix) Corrected error progagation when failing to read sounds and parsing headers.

----------------- Released 1.5b1, 2/16/99 -----------------------

2/21/99 (new feature) Convert command now does low-pass filtering when resampling sounds.

2/21/99 (bug fix) Corrected the code for immediate loading, which would have caused problems in future versions.

2/21/99 (new feature) Improved documentation and configuration of the snackSphere package.

2/22/99 (bug fix) Corrected right edge padding of spectrograms.

2/28/99 (new feature) Created commands getOGIwave and createOGIwave for interaction between Snack sounds and the wave objects of the OGI toolkit.

3/1/99 (new feature) Added support for reading the MPEG layer 3 sound format.

----------------- Released 1.5b2, 3/3/99 -----------------------

3/17/99 (new feature) Added Snack_GetSound() function which returns a sound struct given a sound command name.

3/18/99 (bug fix) Corrected bug for file based playback of MP3 sounds.

----------------- Released 1.5b3, 3/18/99 -----------------------

3/20/99 (feature change) A number of options to the play and record commands are now deprecated. The configure command provides this functionality. Also, sound command options should now be written in lower case only.

3/23/99 (removed feature) The spectrogram widget was removed as all functionality is provided by the corresponding canvas item.


src/snack.c
/* 
 * Copyright (C) 1997-2000 Kare Sjolander <kare@speech.kth.se>
 *
 * This file is part of the Snack sound extension for Tcl/Tk.
 * The latest version can be found at http://www.speech.kth.se/snack/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "snack.h"
#include "tk.h"
#include "jkCanvItems.h"

#if defined(__WIN32__)
#  define WIN32_LEAN_AND_MEAN
#  include <windows.h>
#  undef WIN32_LEAN_AND_MEAN

BOOL APIENTRY
DllMain(HINSTANCE hInst, DWORD reason, LPVOID reserved)
{
  return TRUE;
}
#endif

extern Tk_ItemType snackWaveType;
extern Tk_ItemType snackSpectrogramType;
extern Tk_ItemType snackSectionType;
extern Tk_CustomOption waveTagsOption;
extern Tk_CustomOption spegTagsOption;
extern Tk_CustomOption sectTagsOption;

#define play_width 19
#define play_height 19
static unsigned char play_bits[] = {
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x18, 0x00, 0x00,
   0x78, 0x00, 0x00, 0xf8, 0x01, 0x00, 0xf8, 0x07, 0x00, 0xf8, 0x1f, 0x00,
   0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf8, 0x1f, 0x00, 0xf8, 0x07, 0x00,
   0xf8, 0x01, 0x00, 0x78, 0x00, 0x00, 0x18, 0x00, 0x00, 0x00, 0x00, 0x00,
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

#define rec_width 19
#define rec_height 19
static unsigned char rec_bits[] = {
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x07, 0x00,
   0xe0, 0x1f, 0x00, 0xf0, 0x3f, 0x00, 0xf0, 0x3f, 0x00, 0xf8, 0x7f, 0x00,
   0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf0, 0x3f, 0x00,
   0xf0, 0x3f, 0x00, 0xe0, 0x1f, 0x00, 0x80, 0x07, 0x00, 0x00, 0x00, 0x00,
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

#define stop_width 19
#define stop_height 19
static unsigned char stop_bits[] = {
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf8, 0x7f, 0x00,
   0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00,
   0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00,
   0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0xf8, 0x7f, 0x00, 0x00, 0x00, 0x00,
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

#define pause_width 19
#define pause_height 19
static unsigned char pause_bits[] = {
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf8, 0x7c, 0x00,
   0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00,
   0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00,
   0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00, 0xf8, 0x7c, 0x00, 0x00, 0x00, 0x00,
   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

extern Tcl_Channel snackDebugChannel;

int
Snack_InitCmd(ClientData cdata, Tcl_Interp *interp, int objc,
	      Tcl_Obj *CONST objv[])
{
  int len;
  char *str;

  if (objc == 2) {
    if (Tcl_IsSafe(interp)) {
      Tcl_AppendResult(interp, "can't open log file in a safe interpreter",
		       (char *) NULL);
      return TCL_ERROR;
    }
    str = Tcl_GetStringFromObj(objv[1], &len);
    snackDebugChannel = Tcl_OpenFileChannel(interp, str, "w", 420);
  }

  return TCL_OK;
}

Tcl_HashTable jkSoundTable;

#ifdef SNACK_CSLU_TOOLKIT
extern int fromCSLUshWaveCmd(Sound *s, Tcl_Interp *interp, int objc,
			     Tcl_Obj *CONST objv[]);
extern int toCSLUshWaveCmd(Sound *s, Tcl_Interp *interp, int objc,
			   Tcl_Obj *CONST objv[]);
#endif

int useOldObjAPI = 0;

extern SnackStubs *snackStubs;

int
Snack_Init(Tcl_Interp *interp)
{
  Tcl_CmdInfo infoPtr;
  char *version;

#ifdef USE_TCL_STUBS
  if (Tcl_InitStubs(interp, "8", 0) == NULL) {
    return TCL_ERROR;
  }
#endif

  version = Tcl_GetVar(interp, "tcl_version",
		       (TCL_GLOBAL_ONLY | TCL_LEAVE_ERR_MSG));
  
  if (strcmp(version, "8.0") == 0) {
    useOldObjAPI = 1;
  }

#ifdef TCL_81_API
  if (Tcl_PkgProvideEx(interp, "snack", SNACK_VERSION,
		       (ClientData) &snackStubs) != TCL_OK) {
    return TCL_ERROR;
  }
#else
  if (Tcl_PkgProvide(interp, "snack", SNACK_VERSION) != TCL_OK) {
    return TCL_ERROR;
  }
#endif

  SnackMixerOpen();

  if (Tcl_GetCommandInfo(interp, "button", &infoPtr) != 0) {
#ifdef USE_TK_STUBS
    if (Tk_InitStubs(interp, "8", 0) == NULL) {
      return TCL_ERROR;
    }
#endif
    Tk_CreateItemType(&snackWaveType);
    Tk_CreateItemType(&snackSpectrogramType);
    Tk_CreateItemType(&snackSectionType);
    Tk_DefineBitmap(interp, Tk_GetUid("play"),   (char *) play_bits,
		    play_width, play_height);
    Tk_DefineBitmap(interp, Tk_GetUid("record"), (char *) rec_bits,
		    rec_width, rec_height);
    Tk_DefineBitmap(interp, Tk_GetUid("stop"),   (char *) stop_bits,
		    stop_width, stop_height);
    Tk_DefineBitmap(interp, Tk_GetUid("pause"),  (char *) pause_bits,
		    pause_width, pause_height);
    waveTagsOption.parseProc = Tk_CanvasTagsParseProc;
    waveTagsOption.printProc = Tk_CanvasTagsPrintProc;
    spegTagsOption.parseProc = Tk_CanvasTagsParseProc;
    spegTagsOption.printProc = Tk_CanvasTagsPrintProc;
    sectTagsOption.parseProc = Tk_CanvasTagsParseProc;
    sectTagsOption.printProc = Tk_CanvasTagsPrintProc;
  }

  Tcl_CreateObjCommand(interp, "sound", Snack_SoundCmd,
		       NULL, (Tcl_CmdDeleteProc *)NULL);
  
  Tcl_CreateObjCommand(interp, "audio", Snack_AudioCmd,
		       NULL, Snack_AudioDeleteCmd);

  Tcl_CreateObjCommand(interp, "initSnack", 
		       (Tcl_ObjCmdProc*) Snack_InitCmd,
		       NULL, (Tcl_CmdDeleteProc *)NULL);

  snackDebugChannel = Tcl_GetStdChannel(TCL_STDERR);

  Tcl_SetVar(interp, "snack_patchLevel", SNACK_PATCH_LEVEL, TCL_GLOBAL_ONLY);
  Tcl_SetVar(interp, "snack_version",    SNACK_VERSION,     TCL_GLOBAL_ONLY);
  
  AddSnackNativeFormats();

  Tcl_InitHashTable(&jkSoundTable, TCL_STRING_KEYS);

  Tcl_CreateExitHandler(Snack_ExitProc, (ClientData) NULL);

#ifdef SNACK_CSLU_TOOLKIT
  Snack_AddSubCmd(SNACK_SOUND_CMD, "fromCSLUshWave",
		  (Snack_CmdProc *) fromCSLUshWaveCmd, NULL);
  Snack_AddSubCmd(SNACK_SOUND_CMD, "toCSLUshWave",
		  (Snack_CmdProc *) toCSLUshWaveCmd, NULL);
#endif

  return TCL_OK;
}

int
Snack_SafeInit(Tcl_Interp *interp)
{
  return Snack_Init(interp);
}

/*
static double _t0;

void
TimerStart() {
  _t0 = SnackCurrentTime();
}

void
TimerStop() {
  char buf[20];
  sprintf(buf, "%f", SnackCurrentTime()-_t0);
  Tcl_Write(snackDebugChannel, buf, strlen(buf));
  Tcl_Write(snackDebugChannel, "\n", 1);
  Tcl_Flush(snackDebugChannel);
}
*/
